### PR TITLE
Feature/odlc image processing

### LINKED ISSUE #7: ODLC Image Processing
# ODLC Image Processing

## Problem

Given an image, process it to make it suitable for contour detection.

## Solution

Given an image, process it to make it suitable for contour detection.

### PR DESCRIPTION
**Pull request by fallscameron01**
closes #7 

# Feature/odlc image processing

Implements image preprocessing for use in contour detection of standard odlc objects.

**Changes**

- add `odlc_image_processing.py`
- add `preprocess_std_odlc()` function that runs preprocessing for the image

### CODE REVIEW

**Comment by fallscameron01** [comment]
Example of preprocessed image:
Preprocessed Image

**Comment by eblake003** [comment]
My best idea right now is to make the edge detection thresholds a parameter, then iterate through a bunch of different inputs and run the ODLC shape filtering and categorization on each one. This hopefully won't have a huge impact, and won't require as much fine-tuning of parameters.

File: vision/common/constants.py [hunk 1]
```diff
@@ -9,6 +9,9 @@
 # single channel image of booleans
 Mask: TypeAlias = NDArray[Shape["*, *"], Bool]
 
+# A kernel for image processing steps
+Kernel: TypeAlias = NDArray[Shape["*, *"], UInt8]
```
Comment (fallscameron01) [inline vision/common/constants.py]: Does this need to be a type alias? By abstracting like this, you are making the shape arbitrary, which reduces specificity. A 3x3 kernel will be of shape "3, 3", but having a generic `Kernel` type removes that detail.

File: vision/standard_object/odlc_image_processing.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    Preprocesses image for use in detecting the contours of standard objects.
+
+    Parameters
+    ----------
+    image : Image
+        image from airdrop area flyover before any processing has occured
+    thresh_min: int
+        The minimum threshold input for the Canny edge detection
+    thresh_max: int
+        The maximum threshold input for the Canny edge detection
```
Comment (fallscameron01) [inline vision/standard_object/odlc_image_processing.py]: These 2 parameters should have default values so there is a clear idea of the kind of values that this function would be expecting.

File: vision/standard_object/odlc_image_processing.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        the image after preprocessing for use in contour detection/processing
+    """
+    grayscaled: Image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)  # Convert to graycsale
+
+    blurred: Image = cv2.GaussianBlur(grayscaled, ksize=(3, 3), sigmaX=1.5)  # Blur the image
+
+    edges: Image = cv2.Canny(image=blurred, threshold1=thresh_min, threshold2=thresh_max)
+
+    # Create the kernel for the dilation
+    kernel: Kernel = np.ones((3, 3), np.uint8)
```
Comment (fallscameron01) [inline vision/standard_object/odlc_image_processing.py]: See comment above about typing for kernels.

File: vision/standard_object/odlc_image_processing.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    dilated: Image = cv2.dilate(edges, kernel, iterations=1)
```
Comment (fallscameron01) [inline vision/standard_object/odlc_image_processing.py]: This is still a binarized image, correct? Not saying it isn't, just making sure you check that.

Comment (fallscameron01) [inline vision/standard_object/odlc_image_processing.py]: Also, does the dilation have a negative impact on the clarity of the shape?

**Review by fallscameron01 — COMMENTED** [review]
Won't let select "Request changes", but I am requesting changes.

Just a couple of small changes here.

File: vision/standard_object/odlc_image_processing.py (hunk 4 above)
Comment (eblake003) [reply vision/standard_object/odlc_image_processing.py]: The short answer is no, a small dilation like this shouldn't significantly impact the appearance of the shape - unless there are a bunch of edges around it like in the image you sent earlier.

I did have some success getting rid of most of these extra edges by tweaking the parameters, so hopefully the iterative process I am planning on doing in the pipeline will result in one set of parameters working really well. In that case, a dilation should be helpful.

Comment (eblake003) [reply vision/standard_object/odlc_image_processing.py]: So apparently it's not _really_ binarized, it just only has values of 0 or 255. The OpenCV documentation states that all nonzeros are treated as ones for `cv2.findContours()`, so it effectively performs an internal binarization no matter what we pass in.

**Review by fallscameron01 — COMMENTED** [review]
**Approved.**

Looks good. Nice catch on the single channel image types.

**Review by eblake003 — APPROVED** [review]

**Pull request merged**